integrationtest/backend/tck/src/main/java/org/hibernate/search/integrationtest/backend/tck/work/IndexIndexerLargeDocumentsIT.java

```diff
@@ -0,0 +1,175 @@
+import org.hibernate.search.util.impl.test.annotation.TestForIssue;
+
+import org.junit.Before;
+import org.junit.Rule;
+import org.junit.Test;
+
+import com.google.common.io.Resources;
+import org.awaitility.Awaitility;
+
+/**
+ * Test the {@link IndexIndexer} when indexing many large documents, or a single very large document.
+ * <p>
+ * This is relevant when using AWS authentication in particular,
+ * because large documents cause large requests, which causes a different behavior to compute the content length,
+ * and AWS request signing is sensitive to the content length.
+ */
+@TestForIssue(jiraKey = "HSEARCH-4239")
+public class IndexIndexerLargeDocumentsIT {
+
+	// This must be high, otherwise the bug won't be reproduced.
+	private static final int MANY_LARGE_DOCUMENTS_COUNT = 2000;
+
+	private static final String GREAT_EXPECTATIONS;
+	static {
+		try {
+			GREAT_EXPECTATIONS = Resources.toString(
+					IndexIndexerLargeDocumentsIT.class.getResource( "/great_expectations.txt" ),
+					StandardCharsets.UTF_8
+			);
+		}
+		catch (IOException e) {
+			throw new UncheckedIOException( e );
+		}
+	}
+
+	@Rule
+	public final SearchSetupHelper setupHelper = new SearchSetupHelper();
+
+	private final SimpleMappedIndex<IndexBinding> index = SimpleMappedIndex.of( IndexBinding::new );
+
+	public IndexIndexerLargeDocumentsIT() {
+	}
+
+	@Before
+	public void setup() {
+		setupHelper.start().withIndex( index ).setup();
+	}
+
+	@Test
+	public void manyLargeDocuments_add() {
+		doTest( MANY_LARGE_DOCUMENTS_COUNT, this::largeValue, Operation.ADD );
+	}
+
+	@Test
+	public void manyLargeDocuments_addOrUpdate() {
+		doTest( MANY_LARGE_DOCUMENTS_COUNT, this::largeValue, Operation.ADD_OR_UPDATE );
+	}
+
+	@Test
+	public void oneVeryLargeDocuments_add() {
+		doTest( 1, this::veryLargeValue, Operation.ADD );
+	}
+
+	@Test
+	public void oneVeryLargeDocuments_addOrUpdate() {
+		doTest( 1, this::veryLargeValue, Operation.ADD_OR_UPDATE );
+	}
+
+	private void doTest(int count, IntFunction<String> valueProvider, Operation operation) {
+		assertThatQuery( index.query()
+				.where( f -> f.matchAll() ) )
+				.hasNoHits();
+
+		indexAndWait( count, valueProvider, operation );
+
+		index.createWorkspace().refresh().join();
+
+		assertThatQuery( index.query()
+				.where( f -> f.matchAll() ) )
+				.hasTotalHitCount( count );
+	}
+
+	private void indexAndWait(int count, IntFunction<String> valueProvider, Operation operation) {
+		IndexIndexer indexer = index.createIndexer();
+		CompletableFuture<?>[] tasks = new CompletableFuture<?>[count];
+		for ( int i = 0; i < count; i++ ) {
+			final int id = i;
+			tasks[i] = operation.apply( indexer, referenceProvider( String.valueOf( id ) ),
+					document -> document.addValue( index.binding().content, valueProvider.apply( id ) ) );
+		}
+		CompletableFuture<?> future = CompletableFuture.allOf( tasks );
+		Awaitility.await().until( future::isDone );
```

Comment:
This is just a curiosity. Why are we using the `await` method instead of `Future#get` or `join`?

Reply:
So that if the future never finishes, we are 100% sure the test does finish and throws a nice exception that mentions how much time we waited.

Reply:
IMO if the almost the same feature is provided by the `java.util.concurrent` library we don't need to use an external library. On the other hand, that is not absolutely critical.